src/rt/sched/behaviourcore.h

```diff
@@ -57,74 +57,308 @@ namespace verona::rt
+        uintptr_t new_status_val = old_status_val | STATUS_SLOT_ACTIVE_FLAG;
+        if (status.compare_exchange_strong(
+              old_status_val, new_status_val, std::memory_order_acq_rel))
+          break;
+      }
     }
 
+    /**
+     * Get the behaviour associated with the read-only slot
+     */
     BehaviourCore* get_behaviour()
     {
-      return (BehaviourCore*)status.load(std::memory_order_acquire);
+      assert(is_read_only());
+      assert(behaviour.load(std::memory_order_acquire) != nullptr);
+      return behaviour.load(std::memory_order_acquire);
     }
 
+    /**
+     * Set the behaviour associated with the read-only slot
+     */
     void set_behaviour(BehaviourCore* b)
     {
-      status.store((uintptr_t)b, std::memory_order_release);
+      assert(is_read_only());
+      behaviour.store(b, std::memory_order_release);
+    }
+
+    /**
+     * Return the next slot
+     */
+    Slot* next_slot()
+    {
+      return (
+        Slot*)(status.load(std::memory_order_acquire) & STATUS_NEXT_SLOT_MASK);
+    }
+
+    /**
+     * Returns true if current slot is a writer or a blocked reader,
+     * otherwise returns false
+     */
+    bool set_next_slot_reader(Slot* n)
+    {
+      // Check that the last two bits are zero
+      assert(((uintptr_t)n & ~STATUS_NEXT_SLOT_MASK) == 0);
+      while (true)
+      {
+        assert(next_slot() == nullptr);
+        uintptr_t old_status_val = status.load(std::memory_order_acquire);
+        uintptr_t new_status_val =
+          old_status_val | ((uintptr_t)n) | (STATUS_NEXT_SLOT_READER_FLAG);
+        Logging::cout() << "prev slot is_reader" << is_read_only()
+                        << " curr reader " << this
+                        << "old_status_val: " << old_status_val
+                        << " new_status_val: " << new_status_val
+                        << Logging::endl;
+        if (status.compare_exchange_strong(
+              old_status_val, new_status_val, std::memory_order_acq_rel))
+        {
+          if (is_read_only())
+          {
+            if (
+              (old_status_val & STATUS_SLOT_ACTIVE_FLAG) ==
+              STATUS_SLOT_ACTIVE_FLAG)
+              return false;
+            else
+              return true;
+          }
+          else
+            return true;
+        }
+      }
+    }
+
+    /**
+     * Returns the next behaviour.
+     * True only if the next one in the queue is a writer.
+     */
+    BehaviourCore* next_behaviour()
+    {
+      assert(!is_next_slot_read_only());
+      return (
+        BehaviourCore*)(status.load(std::memory_order_acquire) & STATUS_NEXT_SLOT_MASK);
+    }
+
+    /**
+     * Set the next behaviour
+     */
+    void set_next_slot_writer(BehaviourCore* b)
+    {
+      // Check that the last two bits are zero
+      assert(((uintptr_t)b & ~STATUS_NEXT_SLOT_MASK) == 0);
+      status.store(
+        status.load(std::memory_order_acquire) | ((uintptr_t)b),
+        std::memory_order_release);
```

Comment:
This loses if the previous behaviour was an Active Reader?  That seems like it should be important information.

Reply:
Shouldn't we be detecting this case, and setting next_writer if we observe the Active bit set?

Reply:
For writer, it doesn't matter if the reader is active or blocked. In any case, when reader finishes, it will set the "next_writer" variable.